Python/ceval.c

```diff
@@ -5998,3 +6007,38 @@ void Py_LeaveRecursiveCall(void)
 {
     _Py_LeaveRecursiveCall_inline();
 }
+
+static void improve_missing_self_error(PyThreadState* tstate, PyCodeObject* co, Py_ssize_t nargs) {
+
+    if (nargs + 1 != co->co_argcount) {
+        return;
+    }
+    PyObject *type, *val, *tb;
+    _PyErr_Fetch(tstate, &type, &val, &tb);
+
+    if (co->co_argcount > 0 && co->co_varnames && PyTuple_GET_SIZE(co->co_varnames) > 0) {
+        PyObject* first_arg = PyTuple_GET_ITEM(co->co_varnames, 0);
+        int comp = PyUnicode_CompareWithASCIIString(first_arg, "self");
+        if (comp == 0 || comp == -1) {
+            PyErr_Clear();
+            goto exit;
+        }
+    }
+
+    if (PyUnicode_CheckExact(val)) {
+        PyObject* trailing = PyUnicode_FromString(". Did you forget 'self' in the method definition?");
+        if (trailing == NULL) {
+            PyErr_Clear();
+            goto exit;
+        }
+        PyObject* new_val = PyUnicode_Concat(val, trailing);
+        if (new_val == NULL) {
```

Comment:
`trailing` is leaked here.

It may be simpler to use `PyUnicode Format("%U. Did you ...", val)`. Less lines, less variables.

Reply:
Excellent point! It seems that I tend to complicate my life needlessly :)

Reply:
Done!